File: algorithms/04_java/data_structures/SparseTable.java
public class SparseTable {
    private int[][] st;
    private int[] log;

    public SparseTable(int[] a) {
        int n = a.length;
        log = new int[n + 1];
        for (int i = 2; i <= n; i++) log[i] = log[i / 2] + 1;
        int k = log[n] + 1;
        st = new int[n][k];
        for (int i = 0; i < n; i++) st[i][0] = a[i];
        for (int j = 1; j < k; j++)
            for (int i = 0; i + (1 << j) <= n; i++)
                st[i][j] = Math.min(st[i][j - 1], st[i + (1 << (j - 1))][j - 1]);
    }

    public int queryMin(int l, int r) {
        int j = log[r - l + 1];
        return Math.min(st[l][j], st[r - (1 << j) + 1][j]);
    }

    public static void main(String[] args) {
        int[] a = {5, 2, 8, 1, 9, 3, 7};
        SparseTable st = new SparseTable(a);
        if (st.queryMin(0, 6) != 1) throw new AssertionError("min(0,6) should be 1");
        if (st.queryMin(0, 2) != 2) throw new AssertionError("min(0,2) should be 2");
        if (st.queryMin(3, 5) != 1) throw new AssertionError("min(3,5) should be 1");
        if (st.queryMin(4, 4) != 9) throw new AssertionError("min(4,4) should be 9");
        System.out.println("[Java SparseTable] O(1) range minimum query verified");
    }
}

File: algorithms/07_typescript/data_structures/sparse_table.ts
/**
 * Sparse Table for Range Minimum Query (RMQ)
 * O(n log n) preprocessing, O(1) query time for idempotent operations.
 */

export class SparseTable {
  private st: number[][];
  private logTable: number[];

  constructor(arr: number[]) {
    const n = arr.length;
    const k = Math.floor(Math.log2(n)) + 1;
    this.st = Array.from({ length: n }, () => new Array(k).fill(0));
    this.logTable = new Array(n + 1).fill(0);

    for (let i = 2; i <= n; i++) {
      this.logTable[i] = this.logTable[Math.floor(i / 2)] + 1;
    }

    for (let i = 0; i < n; i++) this.st[i][0] = arr[i];

    for (let j = 1; j < k; j++) {
      for (let i = 0; i + (1 << j) <= n; i++) {
        this.st[i][j] = Math.min(this.st[i][j - 1], this.st[i + (1 << (j - 1))][j - 1]);
      }
    }
  }

  queryMin(l: number, r: number): number {
    const len = r - l + 1;
    const k = this.logTable[len];
    return Math.min(this.st[l][k], this.st[r - (1 << k) + 1][k]);
  }
}

const rmq = new SparseTable([7, 2, 3, 0, 5, 10, 3, 12, 18]);
if (rmq.queryMin(1, 4) !== 0 || rmq.queryMin(4, 7) !== 3) {
  throw new Error("Sparse table query failed");
}
console.log("Sparse Table RMQ verified successfully.");
